packages/components/src/slot-fill/fill.ts

```diff
@@ -7,31 +7,26 @@ import { useContext, useLayoutEffect, useRef } from '@wordpress/element';
  * Internal dependencies
  */
 import SlotFillContext from './context';
-import useSlot from './use-slot';
 import type { FillComponentProps } from './types';
 
 export default function Fill( { name, children }: FillComponentProps ) {
 	const registry = useContext( SlotFillContext );
-	const slot = useSlot( name );
+	const instanceRef = useRef( {} );
+	const childrenRef = useRef( children );
 
-	const ref = useRef( {
-		name,
-		children,
-	} );
+	useLayoutEffect( () => {
+		childrenRef.current = children;
+	}, [ children ] );
 
 	useLayoutEffect( () => {
-		const refValue = ref.current;
-		refValue.name = name;
-		registry.registerFill( name, refValue );
-		return () => registry.unregisterFill( name, refValue );
+		const instance = instanceRef.current;
+		registry.registerFill( name, instance, childrenRef.current );
+		return () => registry.unregisterFill( name, instance );
 	}, [ registry, name ] );
 
 	useLayoutEffect( () => {
-		ref.current.children = children;
-		if ( slot ) {
-			slot.rerender();
-		}
-	}, [ slot, children ] );
```

Comment:
This code used to store the new value of `children` into a `ref.current`. But that's a "silent" mutation that doesn't trigger any listeners. That's why the code had to do a manual `slot.rerender()` call. The new code updates the `children` using the `observableMap.set` method (hidden inside the `updateFill` method). Because the `slot` listens for fills updates via `useObservableValue( fills, name )`, it gets notified and rerenders automatically.